snuba_sdk/query_validation.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence, Set, Tuple, Union
+
+# Import the modules due to sphinx autodoc problems
+# https://github.com/agronholm/sphinx-autodoc-typehints#dealing-with-circular-imports
+from snuba_sdk import query as main
+from snuba_sdk import query_visitors as qvisitors
+from snuba_sdk.column import Column
+from snuba_sdk.conditions import (
+    BooleanCondition,
+    Condition,
+    Op,
+    get_top_level_conditions,
+)
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import Expression
+from snuba_sdk.function import CurriedFunction
+from snuba_sdk.relationships import Join
+from snuba_sdk.schema import Column as ColumnModel
+
+
+class InvalidMatchError(Exception):
+    pass
+
+
+def validate_match(
+    query: main.Query, column_finder: qvisitors.ExpressionSearcher
+) -> None:
+    """
+    Perform validation related to the match clause of the query. Currently that takes two forms:
+    validate that the columns referenced in the query are valid columns based on the match clause,
+    and ensure that the query has the appropriate conditions based on the entities required columns.
+
+    :param query: The query to be validated.
+    :type query: main.Query
+    :param column_finder: An ExpressionSearcher used to find all the columns in the query.
+    :type column_finder: ExpressionSearcher
+    """
+    all_columns = column_finder.visit(query)
+    if isinstance(query.match, main.Query):
+        validate_subquery(query.match, all_columns)
+        validate_match(query.match, column_finder)
+    elif isinstance(query.match, Join):
+        validate_join(query.match, all_columns)
+    else:
+        validate_entity(query.match, all_columns)
+
+    validate_required_columns(query)
+
+
+def validate_subquery(match: main.Query, all_columns: Set[Expression]) -> None:
+    """
+    Validate that the outer query is only referencing columns in the inner query.
+
+    :param match: The inner query of the query being validated.
+    :type match: main.Query
+    :param all_columns: All the columns referenced in the outer query.
+    :type all_columns: Set[Expression]
+    :raises InvalidQueryError
+    """
+    inner_match = set()
+    assert match.select is not None
+    for s in match.select:
+        if isinstance(s, CurriedFunction):
+            inner_match.add(s.alias)
```

Comment:
Because in the outer query you reference function calls using the alias. So when validating that the outer query has correct column names, the column names could match with an alias in the inner query.

```
SELECT xfoo <- alias of the inner query
FROM (
    SELECT foo(x) AS xfoo...
```

Reply:
so then why only CurriedFunction(s)? Why not `Function`s and `Lambda`s?

Reply:
Functions are a subclass of CurriedFunctions in this codebase and there is no support for Lambdas yet.